File: leetCode/design/DesignAddAndSearchWordsDataStructure.py
# https://leetcode.com/problems/design-add-and-search-words-data-structure/
class WordDictionary:
    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.words: Dict[int, Set[str]] = {}

    def addWord(self, word: str) -> None:
        """
        Adds a word into the data structure.
        """
        len_w = len(word)
        value = self.words.get(len_w, set())
        value.add(word)
        self.words[len_w] = value

    def search(self, word: str) -> bool:
        """
        Returns if the word is in the data structure. A word could contain the dot character '.' to represent any one letter.
        """
        len_w = len(word)
        to_find = self.words.get(len_w, set())
        if word in to_find:
            return True
        for it_word in to_find:
            is_equal = True
            for i in range(len_w):
                if it_word[i] != word[i] and word[i] != '.':
                    is_equal = False
                if not is_equal:
                    break
            if is_equal:
                return True
        return False


# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)

File: leetCode/tree/design-add-and-search-words-data-structure.rs
// https://leetcode.com/problems/design-add-and-search-words-data-structure
use std::collections::HashMap;
use std::rc::Rc;
use std::cell::RefCell;
use std::str::Chars;

#[derive(Debug, Clone, Default)]
struct WordDictionary {
    next_nodes: HashMap<char, Rc<RefCell<WordDictionary>>>,
    ends_here: bool,
}


/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl WordDictionary {
    fn new() -> Self {
        Self::default()
    }
    
    fn add_word(&mut self, word: String) {
        self.add_word_chars(word.chars());
    }

    fn add_word_chars(&mut self, mut word: Chars<'_>) {
        let Some(first) = word.next() else {
            self.ends_here = true;
            return;
        };
        let trie_next = self.next_nodes.entry(first).or_default();
        trie_next.borrow_mut().add_word_chars(word);
    }
    
    fn search(&self, word: String) -> bool {
        self.search_chars(word.chars())
    }

    fn search_chars(&self, mut word: Chars<'_>) -> bool {
        //println!("word: {}, is_prefix: {is_prefix}", word.clone().collect::<String>());
        let Some(first) = word.next() else {
            return self.ends_here;
        };
        if first != '.' {
            let Some(trie_next) = self.next_nodes.get(&first) else {
                return false;
            };
            return trie_next.borrow().search_chars(word);
        }

        for next_node in self.next_nodes.values() {
            if next_node.borrow().search_chars(word.clone()) {
                return true;
            }
        }

        false
    }
}

/**
 * Your WordDictionary object will be instantiated and called as such:
 * let obj = WordDictionary::new();
 * obj.add_word(word);
 * let ret_2: bool = obj.search(word);
 */
